src/clusterfuzz/_internal/datastore/data_handler.py

```diff
@@ -1593,45 +1593,82 @@ def get_testcase_variant(testcase_id, job_type):
-  if not fuzz_target:
-    fuzz_target = data_types.FuzzTarget(
-        engine=engine_name, project=project, binary=binary_name)
-    fuzz_target.put()
-
-  job_mapping_key = data_types.fuzz_target_job_key(key_name, job_type)
-  job_mapping = ndb.Key(data_types.FuzzTargetJob, job_mapping_key).get()
-  if job_mapping:
-    job_mapping.last_run = utils.utcnow()
-  else:
-    job_mapping = data_types.FuzzTargetJob(
-        fuzz_target_name=key_name,
-        job=job_type,
-        engine=engine_name,
-        last_run=utils.utcnow())
-  job_mapping.put()
-
   logs.log(
       'Recorded use of fuzz target %s.' % key_name,
       project=project,
       engine=engine_name,
       binary_name=binary_name,
       job_type=job_type)
-  return fuzz_target
+  return result
+
+
+def get_or_create_multi(mapping, data_type):
+  """Gets or creates multiple db entities."""
+  keys = list(mapping.keys())
+  entities = ndb.get_multi([ndb.Key(data_type, key) for key in mapping])
+  entities = dict(zip(keys, entities))
+  new_entities = [
+      data_type(**mapping[key]._asdict())  # pylint: disable=protected-access
+      for key, entity in entities.items()
+      if not entity
+  ]
+  new_entities = ndb.get_multi(ndb.put_multi(new_entities))
+  all_entities = [entity for entity in entities.values() if entity] + (
+      new_entities)
+  return all_entities
+
+
+@retry.wrap(
+    retries=FUZZ_TARGET_UPDATE_FAIL_RETRIES,
+    delay=FUZZ_TARGET_UPDATE_FAIL_DELAY,
+    function='datastore.data_handler.record_fuzz_targets')
+def record_fuzz_targets(engine_name, binaries, job_type):
+  """Record existence of fuzz targets to the DB."""
+  # TODO(metzman): All of this code assumes that fuzzing jobs are behaving
+  # reasonably and won't try to DoS us by putting bogus fuzzers in the db.
+  # This should be changed by limiting the number of fuzz targets saved and
+  # putting an expiration on them.
+  if not binaries:
+    logs.log_error('Expected binaries.')
+    return None
+
+  project = get_project_name(job_type)
+
+  Target = collections.namedtuple('Target', ['engine', 'project', 'binary'])
+
+  mapping = {
+      data_types.fuzz_target_fully_qualified_name(engine_name, project, binary):
+      Target(engine_name, project, binary) for binary in binaries
+  }
+  fuzz_targets = get_or_create_multi(mapping, data_types.FuzzTarget)
+
+  FuzzTargetJob = collections.namedtuple(
+      'FuzzTargetJob', ['fuzz_target_name', 'job', 'engine', 'last_run'])
+
+  time_now = utils.utcnow()
+  mapping = {
+      data_types.fuzz_target_job_key(key_name, job_type): FuzzTargetJob(
+          fuzz_target_name=key_name,
+          job=job_type,
+          engine=engine_name,
+          last_run=time_now) for key_name in mapping
+  }
+
+  jobs = get_or_create_multi(mapping, data_types.FuzzTargetJob)
+  for job in jobs:
+    if not job.last_run:
+      job.last_run = utils.utcnow()
+      ndb.put_multi(jobs)
+
+  return fuzz_targets
```

Comment:
Is this return value used anywhere?

Reply:
Yes. In preprocess.

Reply:
Might be missing something but is this return value actually used? Are you talking about:

```python
def handle_fuzz_no_fuzz_target_selected(output):
  save_fuzz_targets(output)
  # Try again now that there are some fuzz targets.
  utask_preprocess(output.uworker_input.fuzzer_name,
                   output.uworker_input.job_type,
                   output.uworker_input.uworker_env)
```

Reply:
In preprocess, we call record_fuzz_target, this calls record_fuzz_targets under the hood. So we do need to return the value.

Reply:
Ah, missed that!